src/data/gradient_index.cc

```diff
@@ -49,12 +49,8 @@ GHistIndexMatrix::GHistIndexMatrix(DMatrix *p_fmat, bst_bin_t max_bins_per_feat,
   this->columns_ = std::make_unique<common::ColumnMatrix>();
 
   // hessian is empty when hist tree method is used or when dataset is empty
-  if (hess.empty() && !std::isnan(sparse_thresh)) {
```

Comment:
Is removing this condition safe? If so, why?

Reply:
Removing of this condition is required for correct initialization of column matrix for 'approx' method. We don't see any problem with removing this condition. Do you see some potential problem in this case?

Reply:
Hi, the `approx` tree method in its current form doesn't use the column matrix. Unlike `hist`, the `approx` needs to generate the histogram index for every iteration for most of the objective functions (like binary classification), and generating a column matrix is quite expensive for both time and memory so I skipped it here and defined a custom partitioner.

Reply:
I see, thanks.
So, what is your opinion about using the same partitioner both for hist and approx methods?

Reply:
I did notice that for the current PartitionBuilder, there are methods only used by approx method and methods only used by hist. So there is not actually much code reuse (maybe just the LeafPartition method?).

Reply:
You are right, the `LeafPartition` has different realizations for `hist` and `approx` methods. However the most part of the code (`UpdatePostion` and constructor) is used by both method.